File: backend/app/routers/menu.py
"""Menu — `menu_section` / `menu_item` CRUD + reorder (owner / assigned
manager / admin) and the PUBLIC structured menu read. See
docs/API_CONTRACTS.md "Menu (`menu_section`, `menu_item`)".

Its own router (backend/CLAUDE.md "one file per resource"), nested under
`/locations/{location_id}` like `deals.py`. Free-tier feature: no `is_paid`
check anywhere except the (currently switched-off) item-photo platform flag
inside `menu_service`.

Route order note: the literal `.../sections/order` and `.../items/order`
PUT routes are distinct from the `{section_id}` / `{item_id}` PATCH/DELETE
routes by HTTP method, so there is no path-parameter shadowing.
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies.auth import (
    CurrentUser,
    get_current_user_optional,
    require_location_write_access,
)
from app.dependencies.db import get_db
from app.schemas.menu import (
    MenuItemCreate,
    MenuItemOrder,
    MenuItemOut,
    MenuItemUpdate,
    MenuOut,
    MenuPhotoAttach,
    MenuPhotoUploadUrlRequest,
    MenuSectionCreate,
    MenuSectionOrder,
    MenuSectionOut,
    MenuSectionUpdate,
)
from app.schemas.photo import UploadUrlResponse
from app.services import menu_service

router = APIRouter(prefix="/locations", tags=["menu"])


@router.get("/{location_id}/menu", response_model=MenuOut)
async def get_location_menu(
    location_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser | None = Depends(get_current_user_optional),
) -> MenuOut:
    """PUBLIC (no auth required) — the whole structured menu. The optional
    caller only matters for the hidden-location rule (an owner/admin/
    assigned manager can still open their own non-active location's menu,
    exactly like `GET /locations/{id}`)."""
    return await menu_service.get_menu(db, location_id, current_user)


# -- sections ---------------------------------------------------------------


@router.post(
    "/{location_id}/menu/sections",
    response_model=MenuSectionOut,
    status_code=status.HTTP_201_CREATED,
)
async def create_menu_section(
    location_id: int,
    body: MenuSectionCreate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> MenuSectionOut:
    return await menu_service.create_section(db, location_id, body, current_user)


@router.put("/{location_id}/menu/sections/order", response_model=MenuOut)
async def reorder_menu_sections(
    location_id: int,
    body: MenuSectionOrder,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> MenuOut:
    return await menu_service.reorder_sections(db, location_id, body, current_user)


@router.patch("/{location_id}/menu/sections/{section_id}", response_model=MenuSectionOut)
async def update_menu_section(
    location_id: int,
    section_id: int,
    body: MenuSectionUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> MenuSectionOut:
    return await menu_service.update_section(db, location_id, section_id, body, current_user)


@router.delete(
    "/{location_id}/menu/sections/{section_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_model=None,
)
async def delete_menu_section(
    location_id: int,
    section_id: int,
    delete_items: bool = Query(default=False),
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> None:
    """Default keeps the group's items (moved to ungrouped);
    `?delete_items=true` deletes them with the group."""
    await menu_service.delete_section(
        db, location_id, section_id, current_user, delete_items=delete_items
    )


# -- items ------------------------------------------------------------------


@router.post(
    "/{location_id}/menu/items",
    response_model=MenuItemOut,
    status_code=status.HTTP_201_CREATED,
)
async def create_menu_item(
    location_id: int,
    body: MenuItemCreate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> MenuItemOut:
    return await menu_service.create_item(db, location_id, body, current_user)


@router.put("/{location_id}/menu/items/order", response_model=MenuOut)
async def reorder_menu_items(
    location_id: int,
    body: MenuItemOrder,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> MenuOut:
    return await menu_service.reorder_items(db, location_id, body, current_user)


@router.patch("/{location_id}/menu/items/{item_id}", response_model=MenuItemOut)
async def update_menu_item(
    location_id: int,
    item_id: int,
    body: MenuItemUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> MenuItemOut:
    return await menu_service.update_item(db, location_id, item_id, body, current_user)


@router.delete(
    "/{location_id}/menu/items/{item_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_model=None,
)
async def delete_menu_item(
    location_id: int,
    item_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> None:
    await menu_service.delete_item(db, location_id, item_id, current_user)


# -- item photos (built, switched OFF behind the platform flag) --------------


@router.post("/{location_id}/menu/photo-upload-url", response_model=UploadUrlResponse)
async def create_menu_photo_upload_url(
    location_id: int,
    body: MenuPhotoUploadUrlRequest,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> UploadUrlResponse:
    return await menu_service.create_photo_upload_url(db, location_id, body.content_type)


@router.put("/{location_id}/menu/items/{item_id}/photo", response_model=MenuItemOut)
async def set_menu_item_photo(
    location_id: int,
    item_id: int,
    body: MenuPhotoAttach,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> MenuItemOut:
    return await menu_service.set_item_photo(db, location_id, item_id, body, current_user)


@router.delete("/{location_id}/menu/items/{item_id}/photo", response_model=MenuItemOut)
async def remove_menu_item_photo(
    location_id: int,
    item_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> MenuItemOut:
    return await menu_service.remove_item_photo(db, location_id, item_id, current_user)

File: frontend/src/lib/api/menu.ts
// Typed client for the menu sub-resource of `/locations/{id}` —
// docs/API_CONTRACTS.md "Menu (`menu_section`, `menu_item`)". The GET is
// PUBLIC (free-tier, never gated on `is_paid`); every write is owner/
// assigned-manager/admin (checked server-side on each call).
import { apiFetch } from "./client";
import type {
  CreateMenuItemInput,
  CreateMenuSectionInput,
  MenuItem,
  MenuResponse,
  MenuSection,
  UpdateMenuItemInput,
  UpdateMenuSectionInput,
} from "@/types/menu";
import type { PhotoUploadUrlResponse } from "@/types/location";

/**
 * GET /locations/{id}/menu — public. Like `getLocationById`, pass
 * `accessToken` when the caller is signed in and may need to see their OWN
 * non-active location's menu (the editor, an owner previewing); omit it for
 * a genuinely public/SSR read (cached 60s).
 */
export async function getLocationMenu(
  locationId: number,
  accessToken?: string
): Promise<MenuResponse> {
  return apiFetch<MenuResponse>(
    `/locations/${locationId}/menu`,
    { method: "GET" },
    { accessToken, revalidateSeconds: accessToken ? undefined : 60 }
  );
}

/** POST /locations/{id}/menu/sections. */
export async function createMenuSection(
  locationId: number,
  input: CreateMenuSectionInput,
  accessToken: string
): Promise<MenuSection> {
  return apiFetch<MenuSection>(
    `/locations/${locationId}/menu/sections`,
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}

/** PATCH /locations/{id}/menu/sections/{section_id}. */
export async function updateMenuSection(
  locationId: number,
  sectionId: number,
  input: UpdateMenuSectionInput,
  accessToken: string
): Promise<MenuSection> {
  return apiFetch<MenuSection>(
    `/locations/${locationId}/menu/sections/${sectionId}`,
    { method: "PATCH", body: JSON.stringify(input) },
    { accessToken }
  );
}

/**
 * DELETE /locations/{id}/menu/sections/{section_id}. By default the group's
 * items are KEPT and moved to ungrouped; `deleteItems: true` deletes them
 * with the group.
 */
export async function deleteMenuSection(
  locationId: number,
  sectionId: number,
  deleteItems: boolean,
  accessToken: string
): Promise<void> {
  const query = deleteItems ? "?delete_items=true" : "";
  return apiFetch<void>(
    `/locations/${locationId}/menu/sections/${sectionId}${query}`,
    { method: "DELETE" },
    { accessToken }
  );
}

/** PUT /locations/{id}/menu/sections/order — the FULL id list, new order. */
export async function reorderMenuSections(
  locationId: number,
  ids: number[],
  accessToken: string
): Promise<MenuResponse> {
  return apiFetch<MenuResponse>(
    `/locations/${locationId}/menu/sections/order`,
    { method: "PUT", body: JSON.stringify({ ids }) },
    { accessToken }
  );
}

/** POST /locations/{id}/menu/items. */
export async function createMenuItem(
  locationId: number,
  input: CreateMenuItemInput,
  accessToken: string
): Promise<MenuItem> {
  return apiFetch<MenuItem>(
    `/locations/${locationId}/menu/items`,
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}

/** PATCH /locations/{id}/menu/items/{item_id}. */
export async function updateMenuItem(
  locationId: number,
  itemId: number,
  input: UpdateMenuItemInput,
  accessToken: string
): Promise<MenuItem> {
  return apiFetch<MenuItem>(
    `/locations/${locationId}/menu/items/${itemId}`,
    { method: "PATCH", body: JSON.stringify(input) },
    { accessToken }
  );
}

/** DELETE /locations/{id}/menu/items/{item_id}. */
export async function deleteMenuItem(
  locationId: number,
  itemId: number,
  accessToken: string
): Promise<void> {
  return apiFetch<void>(
    `/locations/${locationId}/menu/items/${itemId}`,
    { method: "DELETE" },
    { accessToken }
  );
}

/** PUT /locations/{id}/menu/items/order — the FULL id list of one group
 * (`sectionId` null = the ungrouped list), new order. */
export async function reorderMenuItems(
  locationId: number,
  sectionId: number | null,
  ids: number[],
  accessToken: string
): Promise<MenuResponse> {
  return apiFetch<MenuResponse>(
    `/locations/${locationId}/menu/items/order`,
    { method: "PUT", body: JSON.stringify({ section_id: sectionId, ids }) },
    { accessToken }
  );
}

/** POST /locations/{id}/menu/photo-upload-url — presigned S3 POST for a menu
 * photo (JPEG/PNG, 2MB). 403 `menu_photos_disabled` while the platform flag
 * is off. */
export async function getMenuPhotoUploadUrl(
  locationId: number,
  contentType: string,
  accessToken: string
): Promise<PhotoUploadUrlResponse> {
  return apiFetch<PhotoUploadUrlResponse>(
    `/locations/${locationId}/menu/photo-upload-url`,
    { method: "POST", body: JSON.stringify({ content_type: contentType }) },
    { accessToken }
  );
}

/** PUT /locations/{id}/menu/items/{item_id}/photo — attach or replace. */
export async function setMenuItemPhoto(
  locationId: number,
  itemId: number,
  s3Key: string,
  accessToken: string
): Promise<MenuItem> {
  return apiFetch<MenuItem>(
    `/locations/${locationId}/menu/items/${itemId}/photo`,
    { method: "PUT", body: JSON.stringify({ s3_key: s3Key }) },
    { accessToken }
  );
}

/** DELETE /locations/{id}/menu/items/{item_id}/photo. */
export async function removeMenuItemPhoto(
  locationId: number,
  itemId: number,
  accessToken: string
): Promise<MenuItem> {
  return apiFetch<MenuItem>(
    `/locations/${locationId}/menu/items/${itemId}/photo`,
    { method: "DELETE" },
    { accessToken }
  );
}
